wikidatarefisland/pipes/item_extractor_pipe.py

```diff
@@ -0,0 +1,78 @@
+from itertools import chain
+from .abstract_pipe import AbstractPipe
+from wikidatarefisland.data_model import StatementFilterer
+from ..data_model.filters import StatementFilters
+
+
+class ItemExtractorPipe(AbstractPipe):
+    """Pipe that extracts statements that could be referenced from an item"""
+    def __init__(self, external_id_formatter,
+                 blacklisted_properties=None, whitelisted_ext_ids=None):
+        self.external_id_formatter = external_id_formatter
+        if blacklisted_properties is None:
+            blacklisted_properties = []
+        self.blacklisted_properties = blacklisted_properties
+        if whitelisted_ext_ids is None:
+            whitelisted_ext_ids = []
+        self.whitelisted_ext_ids = whitelisted_ext_ids
+
+    def flow(self, input_data):
+        return self._process_item(input_data)
+
+    def _process_item(self, item_data):
+        # TODO: Filter out bail early if item is in excluded list. see: T251282
+        all_statements = list(chain.from_iterable([i for i in item_data['claims'].values()]))
```

Comment:
Can't wait to hear if you found out the reason for the quirkiness around this. Good work!

Reply:
We didn't quite get there but I talked to Amir and we decided a likely case was that due to the lazy evaluation the state of the external_id_formatter could perhaps be changing.

I think the answer (which now sounds obvious) is that this lazy-evaluating function style was a bad idea as soon as it became apparently that the external_id_formatter could change state (e.g. the lookup methods (rather than the constructor) involve both talking to the internet and reading from a cache)

Reply:
Sounds complex, but also a reason for a pastime experiment... thanks for the explenation